// @concordium/web-sdk/lib/esm/types/CcdAmount.js
import { Big } from 'big.js';
import { TypedJsonDiscriminator, makeFromTypedJson } from './util.js';
const MICRO_CCD_PER_CCD = 1000000;
/**
 * The {@linkcode TypedJsonDiscriminator} discriminator associated with {@linkcode Type} type.
 * @deprecated
 */
export const JSON_DISCRIMINATOR = TypedJsonDiscriminator.CcdAmount;
/**
 * Representation of a CCD amount.
 * The base unit of CCD is micro CCD, which is the representation
 * used on chain.
 */
class CcdAmount {
    constructor(
    /** Internal representation of Ccd amount in micro Ccd. */
    microCcdAmount) {
        this.microCcdAmount = microCcdAmount;
        /** Having a private field prevents similar structured objects to be considered the same type (similar to nominal typing). */
        this.__type = JSON_DISCRIMINATOR;
    }
    /**
     * Get a string representation of the CCD amount.
     * @returns {string} The string representation.
     */
    toString() {
        const microCcdString = this.microCcdAmount.toString();
        const padded = microCcdString.padStart(7, '0');
        return `${padded.slice(0, -6)}.${padded.slice(-6)}`;
    }
    /**
     * Get a JSON-serializable representation of the CCD amount in micro CCD.
     * @returns {string} The JSON-serializable representation.
     */
    toJSON() {
        return this.microCcdAmount.toString();
    }
}
/**
 * Converts a `bigint` to a CCD amount in micro CCD.
 * @param {string} json The JSON representation of the CCD amount.
 * @returns {CcdAmount} The CCD amount.
 */
export function fromJSON(json) {
    return fromMicroCcd(json);
}
/**
 * Type predicate for {@linkcode Type}
 *
 * @param value value to check.
 * @returns whether `value` is of type {@linkcode Type}
 */
export function instanceOf(value) {
    return value instanceof CcdAmount;
}
/**
 * Constructs a CcdAmount and checks that it is valid. It accepts a number, string, big, or bigint as parameter.
 * It can accept a string as parameter with either a comma or a dot as the decimal separator.
 *
 * @param microCcdAmount The amount of micro CCD as a number, string, big, or bigint.
 * @throws If an invalid micro CCD amount is passed, i.e. any value which is not an unsigned 64-bit integer
 */
export function fromMicroCcd(microCcdAmount) {
    // If the input is a "BigSource" assert that the number is whole
    if (typeof microCcdAmount !== 'bigint') {
        microCcdAmount = newBig(microCcdAmount);
        if (!microCcdAmount.mod(Big(1)).eq(Big(0))) {
            throw Error('Can not create CcdAmount from a non-whole number!');
        }
        microCcdAmount = BigInt(microCcdAmount.toFixed());
    }
    if (microCcdAmount < 0n) {
        throw new Error('A micro CCD amount must be a non-negative integer but was: ' + microCcdAmount);
    }
    else if (microCcdAmount > 18446744073709551615n) {
        throw new Error('A micro CCD amount must be representable as an unsigned 64 bit integer but was: ' + microCcdAmount);
    }
    return new CcdAmount(microCcdAmount);
}
/**
 * Create a value representing zero CCD.
 * @returns {CcdAmount} A zero amount of CCD
 */
export function zero() {
    return new CcdAmount(0n);
}
/**
 * Creates a CcdAmount from a number, string, big, or bigint.
 *
 * @param ccd The amount of CCD as a number, string, big or bigint.
 * @returns The CcdAmount object derived from the ccd input parameter
 * @throws If a number is passed with several decimal seperators
 * @throws If a negative amount of micro CCD is passed
 * @throws If the micro CCD passed is greater than what can be contained in a 64-bit integer
 */
export function fromCcd(ccd) {
    if (typeof ccd === 'bigint') {
        ccd = ccd.toString();
    }
    const microCcd = newBig(ccd).mul(Big(MICRO_CCD_PER_CCD));
    return fromMicroCcd(microCcd);
}
function newBig(bigSource) {
    if (typeof bigSource === 'string') {
        return Big(bigSource.replace(',', '.'));
    }
    return Big(bigSource);
}
/**
 * Returns the amount of micro CCD as a Big.
 * @param {CcdAmount} amount The CCD amount.
 * @returns {Big} The amount in micro CCD.
 */
export function toMicroCcd(amount) {
    return Big(amount.microCcdAmount.toString());
}
/**
 * Returns the amount of CCD as a Big.
 * @param {CcdAmount} amount The CCD amount.
 * @returns The amount of CCD as a Big
 */
export function toCcd(amount) {
    return Big(amount.microCcdAmount.toString()).div(Big(MICRO_CCD_PER_CCD));
}
/**
 * Converts an amount of CCD to micro CCD and asserts that the amount is a valid amount of CCD.
 *
 * @param ccd The amount of CCD as a number, string, big or bigint.
 * @returns The amount of micro CCD as a Big
 * @throws If a number is passed with several decimal seperators
 * @throws If a negative amount of micro CCD is passed
 * @throws If the micro CCD passed is greater than what can be contained in a 64-bit integer
 */
export function ccdToMicroCcd(ccd) {
    return toMicroCcd(fromCcd(ccd));
}
/**
 * Converts an amount of micro CCD to CCD and asserts that the amount is a valid amount of CCD.
 *
 * @param microCcd The amount of micro CCD as a number, string, big or bigint.
 * @returns The amount of CCD as a Big
 * @throws If a number is passed with several decimal seperators
 * @throws If a negative amount of micro CCD is passed
 * @throws If the micro CCD passed is greater than what can be contained in a 64-bit integer
 */
export function microCcdToCcd(microCcd) {
    return toCcd(fromMicroCcd(microCcd));
}
/**
 * Get CCD amount in the JSON format used when serializing using a smart contract schema type.
 * @param {CcdAmount} amount The amount.
 * @returns {SchemaValue} The schema value representation.
 */
export function toSchemaValue(amount) {
    return amount.microCcdAmount.toString();
}
/**
 * Convert to CCD amount from JSON format used when serializing using a smart contract schema type.
 * @param {SchemaValue} microCcdString The amount in schema format.
 * @returns {CcdAmount} The amount.
 */
export function fromSchemaValue(microCcdString) {
    return new CcdAmount(BigInt(microCcdString));
}
/**
 * Convert an amount of CCD from its protobuf encoding.
 * @param {Proto.Amount} amount The energy in protobuf.
 * @returns {CcdAmount} The energy.
 */
export function fromProto(amount) {
    return new CcdAmount(amount.value);
}
/**
 * Convert an amount of CCD into its protobuf encoding.
 * @param {CcdAmount} amount The CCD amount.
 * @returns {Proto.Amount} The protobuf encoding.
 */
export function toProto(amount) {
    return {
        value: amount.microCcdAmount,
    };
}
/**
 * Constructs a {@linkcode Type} from {@linkcode Serializable}.
 * @deprecated Use  the{@linkcode fromJSON} function instead.
 * @param {Serializable} value
 * @returns {Type} The duration.
 */
export function fromSerializable(value) {
    return fromMicroCcd(value);
}
/**
 * Converts {@linkcode Type} into {@linkcode Serializable}
 * @deprecated Use {@linkcode CcdAmount.toJSON} method instead.
 * @param {Type} value
 * @returns {Serializable} The serializable value
 */
export function toSerializable(value) {
    return value.microCcdAmount.toString();
}
/**
 * Takes an {@linkcode Type} and transforms it to a {@linkcode TypedJson} format.
 * @deprecated Use the {@linkcode CcdAmount.toJSON} method instead.
 * @param {Type} value - The account address instance to transform.
 * @returns {TypedJson} The transformed object.
 */
export function toTypedJSON(value) {
    return {
        ['@type']: JSON_DISCRIMINATOR,
        value: toSerializable(value),
    };
}
/**
 * Takes a {@linkcode TypedJson} object and converts it to instance of type {@linkcode Type}.
 * @deprecated Use the {@linkcode fromJSON} function instead.
 * @param {TypedJson} json - The typed JSON to convert.
 * @throws {TypedJsonParseError} - If unexpected JSON string is passed.
 * @returns {Type} The parsed instance.
 */
export const fromTypedJSON = /*#__PURE__*/ makeFromTypedJson(JSON_DISCRIMINATOR, fromSerializable);
