// @concordium/web-sdk/lib/esm/types/EntrypointName.js
import { isAsciiAlphaNumericPunctuation } from '../contractHelpers.js';
import { TypedJsonDiscriminator, makeFromTypedJson } from './util.js';
/**
 * The {@linkcode TypedJsonDiscriminator} discriminator associated with {@linkcode Type} type.
 * @deprecated
 */
export const JSON_DISCRIMINATOR = TypedJsonDiscriminator.EntrypointName;
/**
 * Type representing an entrypoint of a smart contract.
 * @template S Use for using string literals for the type.
 */
class EntrypointName {
    /**
     * @deprecated Use the {@linkcode EntrypointName.toJSON} method instead.
     */
    get serializable() {
        return this.value;
    }
    constructor(
    /** The internal string value of the receive name. */
    value) {
        this.value = value;
        /** Having a private field prevents similar structured objects to be considered the same type (similar to nominal typing). */
        this.__type = JSON_DISCRIMINATOR;
    }
    /**
     * Get a string representation of the entrypoint name.
     * @returns {string} The string representation.
     */
    toString() {
        return this.value;
    }
    /**
     * Get a JSON-serializable representation of the entrypoint name.
     * @returns {string} The JSON-serializable representation.
     */
    toJSON() {
        return this.value;
    }
}
/**
 * Converts a `string` to an entrypoint name.
 * @param {string} json The JSON representation of the entrypoint name.
 * @throws If the string is not a valid entrypoint name.
 * @returns {EntrypointName<string>} The entrypoint name.
 */
export function fromJSON(json) {
    return fromString(json);
}
/**
 * Unwraps {@linkcode Type} value
 * @deprecated Use the {@linkcode EntrypointName.toJSON} method instead.
 * @param value value to unwrap.
 * @returns the unwrapped {@linkcode Serializable} value
 */
export function toUnwrappedJSON(value) {
    return toString(value);
}
/**
 * Type predicate for {@linkcode Type}
 *
 * @param value value to check.
 * @returns whether `value` is of type {@linkcode Type}
 */
export function instanceOf(value) {
    return value instanceof EntrypointName;
}
/**
 * Create a smart contract entrypoint name from a string, ensuring it follows the required format.
 * @param {string} value The string with the entrypoint name.
 * @throws If the provided value is not a valid entrypoint name.
 * @returns {EntrypointName}
 */
export function fromString(value) {
    if (value.length > 99) {
        throw new Error('Invalid EntrypointName: Can be atmost 99 characters long.');
    }
    if (!isAsciiAlphaNumericPunctuation(value)) {
        throw new Error('Invalid EntrypointName: Must only contain ASCII alpha, numeric and punctuation characters.');
    }
    return new EntrypointName(value);
}
/**
 * Create a smart contract entrypoint name from a string, but _without_ ensuring it follows the required format.
 * It is up to the caller to ensure the string is a valid entrypoint name.
 * @param {string} value The string with the entrypoint name.
 * @returns {EntrypointName}
 */
export function fromStringUnchecked(value) {
    return new EntrypointName(value);
}
/**
 * Convert a entrypoint name to a string
 * @deprecated Use the {@linkcode EntrypointName.toString} method instead.
 * @param {EntrypointName} entrypointName The entrypoint name to stringify.
 * @returns {string}
 */
export function toString(entrypointName) {
    return entrypointName.value;
}
/**
 * Takes an {@linkcode Type} and transforms it to a {@linkcode TypedJson} format.
 * @deprecated Use the {@linkcode EntrypointName.toJSON} method instead.
 * @param {Type} value - The account address instance to transform.
 * @returns {TypedJson} The transformed object.
 */
export function toTypedJSON(value) {
    return {
        ['@type']: JSON_DISCRIMINATOR,
        value: toString(value),
    };
}
/**
 * Takes a {@linkcode TypedJson} object and converts it to instance of type {@linkcode Type}.
 * @deprecated Use the {@linkcode fromJSON} function instead.
 * @param {TypedJson} json - The typed JSON to convert.
 * @throws {TypedJsonParseError} - If unexpected JSON string is passed.
 * @returns {Type} The parsed instance.
 */
export const fromTypedJSON = /*#__PURE__*/ makeFromTypedJson(JSON_DISCRIMINATOR, fromString);
